torch/_dynamo/backends/distributed.py

```diff
@@ -321,14 +358,20 @@ def run_node(self, n: Node) -> Any:
 
             # Finally, we have to produce inputs for use compiling the next submodule,
             # and these need to be FakeTensors, so we execute the module under fake_mode
-            with self.fake_mode:
-                return curr_submod(*new_args, **kwargs)
+            with self.fake_mode, mock.patch.object(
+                self.fake_mode, "allow_non_fake_inputs", True
+            ):
+                if has_tracing_context and invoked_aot_autograd:
+                    return compiled_submod_real(*new_args, **kwargs)
```

Comment:
Do you think we could figure out a way to add some extra assertions and make this even more explicit?

One option:
(1) Manually use the fake mode to convert `*new_args, **kwargs` to be fake
(2) add a loud assert in the runtime code that asserts that whenever that `first_fake_invocation` is set at runtime, all of our tensor args are fake.

Mostly just trying to guard against the error case where a user sends us their runtime inputs, and we accidentally run this new code path and return them fake tensors.

Reply:
> (1) Manually use the fake mode to convert *new_args, **kwargs to be fake
 
 They're fake at this point. I thought I needed the `allow_non_fake` at one point because of a test failure where the parameters weren't being updated to fake. but i just reran tests and didnt get a failure so let me try to remove.

2) sure - can add assertion. in the runtime code do you mean in aot_autograd ?

Reply:
> They're fake at this point.

ah cool


>  in the runtime code do you mean in aot_autograd ?

yep - in the actual aot autograd runtime code (that secretly returns the FakeTensor outputs) - just an assert that all of the "runtime inputs" are also fake would be nice.

Reply:
The parameters are not actually fake, and the outputs might be subclasses. I did add a test that none of the outputs are torch.Tensors though.

Reply:
There’s also an is_fake method in fake_tensor.py that handles subclasses properly (assert all their inner tensors are fake)